testable example docs for training-serialization

ropottnik opened the pull request:
#10 

Updated example from the `Readme.md` to the newest API and converted it into testable docs inside `lib.rs` (actually with _```no_run_ to test if compilation works).

The example for deserialization and encoding was removed from the `Readme.md` since the `lib.rs` already contains a documentation test for that.
 
The `Readme.md` file now contains only references to these testable docs which should prevent them ever to be outdated ever again.

Narsil commented:
Food for thought [cargo-readme](https://github.com/livioribeiro/cargo-readme)

n1t0 commented:
Thank you @ropottnik!

I agree with @Narsil that it is important to keep the examples in the `README.md` as well if possible. The main reason is that the version on `doc.rs` is not necessarily in sync with `master`.

`cargo-readme` seems nice to do this, and we could include it in the CI probably. I guess with the template file we can even keep the banner with the logo.

ropottnik commented:
Makes sense! So would the pipeline then generate the readme on every PR and make an additional commit?

n1t0 commented:
I think I would probably instead make the pipeline generate it, and check that it is similar to what's in the tested commit, failing if not. Just like we do with linters and tests for example.

Do you want to try updating everything with a template and `cargo-readme`? No need to have the CI pipeline ready, (unless you really want to do it), we can add it later.

ropottnik commented:
@n1t0 sure, I'll give it a try.

ropottnik commented:
@n1t0 I added the `README.tpl` to generate the `README.md` as @Narsil suggested. I also added the pipeline step to check equality. a60fb5a and d83c177 show that the pipeline fails on a minor difference and then passes once that diff is taken care of. 

however, it seems as if some pipeline steps regarding `python` fail. this can't be in connection with my PR since no `python` files were touched, right? or am I missing something?

Narsil commented:
No you are right it seems the failing test are not linked to you.

Did you rebase ? It seems the problem lies in what you rebase against but master seems ok so I'm not sure.

ropottnik commented:
Pretty sure I rebased from `master`.

Could you try rerunning the github pipeline on `master` to be absolutely sure that there is a discrepancy in pipeline results? 
The last commit to `master` was 7 days ago and `black` might have changed something since then.

Narsil commented:
Yes it seems black got updated.

You can run
```
pip install -U black
cd bindings/python/
make style
git add ....
```

And see if that fixes it.

ropottnik commented:
I thought fixing the `black` version to what it was in the last master build (`19.10b0`) might be the cleaner option. What do you guys think @Narsil @n1t0 ?

n1t0 commented:
Great thank you @ropottnik!
